src/webots/nodes/utils/WbNodeUtilities.cpp

```diff
@@ -663,7 +663,24 @@ WbMatter *WbNodeUtilities::findUppermostMatter(WbNode *node) {
 }
 
 WbSolid *WbNodeUtilities::findTopSolid(const WbNode *node) {
-  return dynamic_cast<WbSolid *>(const_cast<WbNode *>(findTopNode(node)));
+  if (node == NULL || node->isWorldRoot())
+    return NULL;
+
+  const WbNode *n = node;
+  const WbNode *parent = n->parentNode();
+  WbSolid *topSolid = NULL;
+  WbSolid *currentNode;
+  while (parent) {
+    currentNode = dynamic_cast<WbSolid *>(const_cast<WbNode *>(n));
```

Comment:
I think that it would be nice to declare and initialize `currentNode` on line 674, because you don't use it outside.
Moreover, I would name it `currentSolid`, as it a pointer to a `WbSolid`. (or a `NULL` if it's not a `WbSolid`)